src/main/java/io/fabric8/elasticsearch/plugin/KibanaUserReindexFilter.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.elasticsearch.plugin;
+
+import org.apache.commons.lang.StringUtils;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.netty.handler.codec.http.HttpRequest;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.http.netty.NettyHttpRequest;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestFilter;
+import org.elasticsearch.rest.RestFilterChain;
+import org.elasticsearch.rest.RestRequest;
+
+public class KibanaUserReindexFilter extends RestFilter {
+
+	private final ESLogger logger;
+	private final String proxyUserHeader;
+	private final String kibanaIndex;
+
+	public KibanaUserReindexFilter(final Settings settings, final ESLogger logger){
+		this.logger = logger;
+		this.proxyUserHeader = settings.get(ConfigurationSettings.SEARCHGUARD_AUTHENTICATION_PROXY_HEADER, 
+				ConfigurationSettings.DEFAULT_AUTH_PROXY_HEADER);
+		this.kibanaIndex = settings.get(ConfigurationSettings.KIBANA_CONFIG_INDEX_NAME, 
+				ConfigurationSettings.KIBANA_CONFIG_INDEX_NAME);
+	}
+	
+	@Override
+	public void process(RestRequest request, RestChannel channel, RestFilterChain chain) throws Exception {
+		try {
+			logger.debug("Handling Request in Kibana User Reindex filter...");
+			final String user = getUser(request);
+			final String requestedIndex = getRequestedIndex(request);
+			
+			if ( StringUtils.isNotEmpty(user) && StringUtils.isNotEmpty(requestedIndex) && requestedIndex.equalsIgnoreCase(kibanaIndex)) {
+				String userIndex = requestedIndex + "." + user;
+				logger.debug("Matched for a kibana user header, will set to '{}' for user '{}'", userIndex, user);
+				
+				//update the request URI here
+				request = updateRequestIndex(request, requestedIndex, userIndex);
+				
+				logger.debug("URI for request is '{}' after update", request.uri());
+			}
+
+		} catch (Exception e) {
+			logger.error("Error handling request in OpenShift SearchGuard filter", e);
+		} finally {
+			chain.continueProcessing(request, channel);
+		}
+	}
+
+	private String getUser(RestRequest request) {
+		return StringUtils.defaultIfBlank(request.header(proxyUserHeader), "");
+	}
+	
+	private RestRequest updateRequestIndex(RestRequest request, String oldIndex, String newIndex) {
+		String uri = request.uri();
+		String replaced = uri.replace(oldIndex, newIndex);
```

Comment:
We now pass in the newIndex as ".kibana." + the sha1 hash of the user name since it may contain characters that ES doesn't allow in an index. (line 51)